src/cockpit/channels/filesystem.py

```diff
@@ -190,6 +190,17 @@ def do_data(self, data):
                 raise ChannelError('internal-error',
                                    message=f"Could not find unique file name for replacing {self._path}")
 
+            # if the file exists already and we have an expected tag (so that do_done
+            # fails on conflicts), copy its permissions
+            if self._tag is not None:
+                try:
+                    stat = os.stat(self._path)
```

Comment:
You should call `tag_from_stat()` on this and make sure it matches.

Reply:
`do_done()` already does that. It could additionally do it here, but only with some awkward code duplication wrt. handling absent files and the `-` special value. I updated it accordingly, let's see whether you like it better.